libs/libarchfpga/src/read_xml_arch_file.cpp

```diff
@@ -3179,33 +3200,84 @@ static void ProcessTilePort(pugi::xml_node Node,
     }
 
     CurSite = Parent.first_child();
     while (CurSite) {
         check_node(CurSite, "site", loc_data);
 
-        t_equivalent_site equivalent_site;
-
-        expect_only_attributes(CurSite, {"pb_type"}, loc_data);
+        expect_only_attributes(CurSite, {"pb_type", "priority"}, loc_data);
         /* Load equivalent site name */
-        auto Prop = get_attribute(CurSite, "pb_type", loc_data).value();
-        equivalent_site.pb_type_name = vtr::strdup(Prop);
+        auto Prop = std::string(get_attribute(CurSite, "pb_type", loc_data).value());
+        PhysicalTileType->equivalent_sites_names.push_back(Prop);
+
+        auto LogicalBlockType = get_type_by_name<t_logical_block_type>(Prop.c_str(), LogicalBlockTypes);
+
+        auto priority = get_attribute(CurSite, "priority", loc_data, ReqOpt::OPTIONAL).as_int(0);
+        LogicalBlockType->physical_tiles_priority[priority].push_back(PhysicalTileType->index);
+        PhysicalTileType->logical_blocks_priority[priority].push_back(LogicalBlockType->index);
 
-        PhysicalTileType->equivalent_sites.push_back(equivalent_site);
+        ProcessEquivalentSiteDirects(CurSite, PhysicalTileType, LogicalBlockType, Prop, loc_data);
 
         CurSite = CurSite.next_sibling(CurSite.name());
     }
 }
 
+static void ProcessEquivalentSiteDirects(pugi::xml_node Parent,
+                                         t_physical_tile_type* PhysicalTileType,
+                                         t_logical_block_type* LogicalBlockType,
+                                         std::string site_name,
+                                         const pugiutil::loc_data& loc_data) {
+    pugi::xml_node CurDirect;
+
+    expect_only_children(Parent, {"direct"}, loc_data);
+
+    if (count_children(Parent, "direct", loc_data) < 1) {
+        archfpga_throw(loc_data.filename_c_str(), loc_data.line(Parent),
+                       "There are no direct pin mappings between site %s and tile %s.\n", site_name.c_str(), PhysicalTileType->name);
+    }
+
+    std::unordered_map<int, int> directs_map;
+
+    CurDirect = Parent.first_child();
+    while (CurDirect) {
+        check_node(CurDirect, "direct", loc_data);
+
+        expect_only_attributes(CurDirect, {"from", "to"}, loc_data);
+
+        std::string from, to;
+        // `from` attribute is relative to the physical tile pins
+        from = std::string(get_attribute(CurDirect, "from", loc_data).value());
+
+        // `to` attribute is relative to the logical block pins
+        to = std::string(get_attribute(CurDirect, "to", loc_data).value());
+
+        auto from_pins = ProcessPinString<t_physical_tile_type_ptr>(CurDirect, PhysicalTileType, from.c_str(), loc_data);
+        auto to_pins = ProcessPinString<t_logical_block_type_ptr>(CurDirect, LogicalBlockType, to.c_str(), loc_data);
+
+        // Checking that the number of pins is exactly the same
+        VTR_ASSERT(from_pins.second - from_pins.first == to_pins.second - to_pins.first);
```

Comment:
Should use archfpga_throw over assert here.

Reply:
Done